### PR TITLE
fix #414 Span..ctor with range

### LINKED ISSUE #414: public Span(T[] array, int start, int length) does not check length before assigning
Comments says it throws when 'end index' is not in range, bug length is never checked.
```
/// <exception cref="System.ArgumentOutOfRangeException">
/// Thrown when the specified start or end index is not in range (&lt;0 or &gt;&eq;length).
/// </exception>
```

### PR DESCRIPTION
**Pull request by krwq**

### CODE REVIEW

File: src/System.Slices/src/System/Span.cs [hunk 1]
```diff
@@ -87,6 +87,8 @@ public Span(T[] array, int start, int length)
         {
             Contract.Requires(array != null);
             Contract.RequiresInInclusiveRange(start, array.Length);
+            Contract.RequiresNonNegative(length);
+            Contract.RequiresInInclusiveRange(start + length, array.Length);
             if (start < array.Length)
```
Comment (livarcocc) [inline src/System.Slices/src/System/Span.cs]: Do you still need this check after adding the Contract above?

**Comment by livarcocc** [comment]
LGTM

File: src/System.Slices/src/System/Span.cs (hunk 1 above)
Comment (krwq) [inline src/System.Slices/src/System/Span.cs]: All of them are needed:
`Contract.RequiresInInclusiveRange(start, array.Length);` - case where start can be negative or completely outside of range.
`Contract.RequiresNonNegative(length);` - case where length is negative but both calculated start and end are within acceptable range
`Contract.RequiresInInclusiveRange` - start withing accepotable range, length positive but too big to fit in the array.

What could work is changing first range check with checking if start is positive as the start + length would detect the range problem - I personally prefer to keep them separate as I have a feeling we will end up with different exception message for both of them after design reviews (hopefully compiler will detect this simple double range check anyway and optimize)

Comment (krwq) [inline src/System.Slices/src/System/Span.cs]: Please file an issue on me if you think this should be fixed.

Comment (livarcocc) [inline src/System.Slices/src/System/Span.cs]: I meant the check inside the method, not the Contract checks. I thought that the contract checks are already verifying the if (start < array.Length). If that's not the case, than I have no issues with it.

**Pull request merged**